bip-tx-terminology.mediawiki

```diff
@@ -0,0 +1,291 @@
+|-
+| P2SH-P2WSH || <nowiki>[…Script Arguments, Witness Script]</nowiki>
+|-
+| P2WPKH || rowspan="4" | Witness Program || <nowiki>OP_0 <PKH></nowiki> || rowspan="4" | NULL || <nowiki>[<SIG>, <PK>]</nowiki>
+|-
+| P2WSH || <nowiki>OP_0 <SH></nowiki>  || <nowiki>[…Script Arguments, Witness Script]</nowiki>
+|-
+| P2TR KP || rowspan="2" | <nowiki>OP_1 <PK></nowiki> || <nowiki>[<SIG>]</nowiki>
+|-
+| P2TR SP || <nowiki>[…Script Arguments, Leaf Script, Control Block, (Opt: Annex)]</nowiki>
+|}
+
+
+==Rationale==
+
+; Amount vs Value
+: Amount underscores that we are referring to a count of satoshis rather than subjective value. As Bitcoin transactions (and code altogether) are littered with “values”, we prefer “amount” as being less ambiguous.
+
+; Witness
+: Ambiguous, use Witness Stack to refer to the witness data of a single Transaction Input, use Witness Structure when talking about all witness data of a transaction
+
+; Forwarding vs Condition Scripts
+: We find that we generally distinguish two types of functions in “scripts”. The Condition Scripts encode the actual spending conditions, while Forwarding Scripts just commit the spender to providing another script down the line.  [[File:bip-tx-terminology/scripts.png|framed|center|alt=An overview of forwarding and condition scripts|Figure 1]]
+
+; Bare Output Scripts
+: Include P2PK, P2PKH, P2MS, and arbitrary other condition scripts directly defined in the Output Script
+
+; Flag
+: While the Flag currently only indicates the presence of a Witness Structure, it could perhaps also indicate the presence of other extended serialization formats in the future, therefore, it’s only flag, not “witness flag”
+
+; P2SH Program
+: We introduce the term P2SH Program to refer to the commitment in an Output Script that forwards execution to the input script corresponding to witness programs
+
+; Witness Item vs Witness Element
+: BIP341 refers to parts of the witness stack as "witness elements" and "witness stack elements", but BIP144 speaks of "item count". We prefer Witness Item to refer to parts of a Witness Stack.
+
+; Witness Section vs Witness Structure
+: While Witness Section and Witness Structure both are used, we prefer the use of Witness Structure as introduced in BIP144
+
+; Witness Stack
+: While BIP144 refers to witness stacks as “script witnesses”, they are not scripts. Strictly speaking, they’re also not stacks, because some Witness Items that appear in Witness Stacks are not added to the stack, such as Control Blocks. We prefer Witness Stack as it is well-established.
+
+; Output Script vs Locking Script
+: The scriptPubKey is also sometimes referred to as a "locking script". However, we aim to emphasize the position of the field in the transaction, as it can either take the function of a condition or forwarding script. We therefore prefer a name that references the location in the transaction rather than a function it does not always have.
```

Comment:
First of all, great initiative and great job explaining the terminology. I only have a couple of related suggestions. Rather than commenting in several places I would argue for a couple of changes here, in one place.

I would suggest that Locking Script and Unlocking Script are also mentioned 'aka' to Output Script and Input Script respectively.

I found that a lot of people have been using them this way and I have adopted it with success, i.e. people understood much easier what was the purpose of the script. I also I have never seen condition script and script arguments to be associated with locking and unlocking script respectively, but maybe I just never happened to come by those!

Rationale
I agree with emphasizing the position of the field. Output Script (as well as Input Script) should be the main entry.

However, an output script is what effectively locks the funds and I argue that it always has this function. e.g. P2SH locks the funds to an unknown script that will be revealed later (similarly for taproot alternative paths/scripts).

Similarly an input script is always what unlocks the funds. It includes everything required to unlock them (incl.any forwarding scripts, etc.).

Suggested changes:
Input Script entry: include "aka scriptSig or Unlocking Script"
Output Script entry: include "aka scriptPubKey or Locking Script". 
Condition Script entry: just remove "aka locking script" 
Script Arguments entry: just remove "aka Unlocking Script"
Rephrase (or maybe remove) this section.

Thanks again for your efforts.

Reply:
Hi @karask, thanks for your comment. I should clarify that _condition script_ and _script arguments_ are newly introduced terms per this proposal specifically to decouple the position of the input script and output script from the function of locking and unlocking, and to underscore the evolution of output types from executable scripts to templates with additional meaning.

As output types have evolved over the years, we have departed from them relying on fully specified executable scripts, but rather imbued certain templates with additional constraints. For example a _P2SH program_ as written only checks that the _redeem script_ revealed in the _input script_ matches the pre-image in the _output script_, but the implied meaning of the _P2SH program_ additionally requires the _redeem script_ to be evaluated satisfactorily. Later with wrapped P2WSH, we don’t even require the witness script to contain all the arguments with push operations directly the script but rather provide them as separate items akin to a pre-built stack. These stand-alone _witness items_ is what I refer to as _script arguments_. With native segwit outputs finally, the _input script_ is empty altogether, and it would feel weird to me to refer to still refer to it as “unlocking script” when it holds no sway in the authorization of spending a UTXO at all anymore.

I think I’ve identified a few points that I need to rework, and I will look to incorporate my explanation into the rationale as well as the descriptions of _input script_ and _output script_. I’ll think about how to best mention “locking script” and “unlocking script”. I’ll also try to rework my description of _forwarding scripts_ to honor their function in committing the spender to the _condition script_ even when they no longer directly express the spending conditions.

Reply:
Murch - Are Forwarding Scripts (Witness Programs and P2SH Programs) not considered to be locking scripts?

Is that why "locking script" is specifically associated with Condition Scripts and not general Output Scripts, as karask suggests?

I ask because I agree with the first two suggestions of this comment, but also am worried taking that position is reinforcing confusion of the terms "locking script" and "unlocking script". I second karask's point that locking script and unlocking script are comparatively easier to understand, and I feel like they have seen a lot of success. Maybe because  they naturally invoke an image of a lock and key. 

They definitely should be incorporated in this BIP, as you have done, though the devil is in the details of how they should be defined/represented. I think what I'm trying to say is the terms locking/unlocking script will likely be easy and accessible entry points for people trying to understand this doc and level up their transaction terminology so care must be taken in how they are represented! :smile:

Reply:
I just realized while I was typing my comment I missed your response. Please disregard if I've said anything that has already been answered :see_no_evil:

Reply:
That’s actually a good point. Perhaps “forwarding scripts” and “condition scripts” are both subtypes of “locking scripts”.

Reply:
@Xekyo thanks for the clarification; that's why I've never heard of them :-)
I have some comments on that too but I will give it some time to wrap my mind around those...

I want to emphasize that regardless of what we say here output script and input script as locking and unlocking scripts are already quite widely used (i.e. as synonyms). I don't think we should conflate their meaning with new concepts; introducing new concepts (e.g. condition script?) would not confuse though.

Reply:
Added as a todo to elaborate on

Reply:
@karask: I now mention “locking script” as a synonym of _output script_, and “unlocking script” as synonym for _input script_. I also rewrote the rationale sections regarding them. They could probably still use some polish, but I’m starting to suffer from tunnel vision—I think I’ve rewritten it four times meanwhile. :grimacing:

I could use a set of fresh eyes to tell me whether it makes any sense, if you have a moment to glance at those definitions and rationales specifically. :)